File: LeetCode 34 -  Find First and Last Position of Element in Sorted Array.java
// Time Complexity: Since we use two binary search methods here: 
// one is to search the 
// leftmost, the other is to search the rightmost. 
// So, the time complexity will be 2*O(logn) --> O(logn)
/* --------------------------------------------------------*/
// Space Complexity: Even though we create new array to present the final,
// however the size is only 2, the space complexity will be O(1).
/* --------------------------------------------------------*/
// I am still confused about the how to consider the boundaries of these kinds 
// of problems, so that we will not miss any ones?

class Solution {
    
    public int[] searchRange(int[] nums, int target) {
        int [] res = new int[]{-1, -1};
        if(nums == null || nums.length == 0) return res;
        res[0] = findLeft(nums, target);
        res[1] = findRight(nums, target);
        return res;
    }
    public int findLeft(int[] nums, int target){
        int idx = -1;
        int low = 0, high = nums.length - 1;
        while(low <= high){
            int mid = low + (high - low)/2;
            // find the leftmost index
            if(nums[mid] >= target){
                high = mid - 1;
            }else{
                low = mid + 1;
            }
            // update the index of target
            if(nums[mid] == target) idx = mid;
        }
        return idx;
    }
    public int findRight(int[] nums, int target){
        int idx = -1;
        int low = 0, high = nums.length - 1;
        while(low <= high){
            int mid = low + (high - low)/2;
            // find the rightmost index
            if(nums[mid] <= target){
                low = mid + 1;
            }else{
                high = mid - 1;
            }
            // update the index of target
            if(nums[mid] == target) idx = mid;
        }
        return idx;
    }
}

File: Leetcode-34_Find First and Last Position of Element in Sorted Array.cpp
// Time Complexity : O(LogN)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

class Solution 
{
    public:
        vector<int> searchRange(vector<int>& nums, int target) 
        {
            vector<int>res(2,-1);
            
            int low = 0, high = nums.size()-1;
            
            if (high == -1 || (high==0 && nums[high] != target))
                return res;
            
            res[0] = returnFirstOrLastIndex(nums, low, target);
            
            if(res[0] == -1)
                return res;
            
            res[1] = returnFirstOrLastIndex(nums, res[0], target, 0);
            
        return res;
                
        }
    
    int returnFirstOrLastIndex(vector<int> & nums, int low, int target, int first=1)
    {
        int high = nums.size()-1;
        
        while(low<=high)
        {
            int mid = low+ (high-low)/2;
            
            if(nums[mid] == target)
            {
                if (first ==1)
                {
					// if mid is first, last element, otherwise previous ele is less than target
                    if(mid == low || nums[mid-1] <target)
                        return mid;
                     
					// proceed to left half to find the first occurrence
                    else
                        high = mid-1;
                }
                
                else
                {
					// if mid is first, last element, otherwise next ele is greater than target
                    if(mid ==high || nums[mid+1] >target)
                        return mid;
                      
					// proceed to right half to find the last occurrence					  
                    else
                        low = mid+1;
                }
            }
            
			
			// search in left half
            else if(nums[mid]> target)
                high = mid-1;
            
			// search in right half
            else
                low = mid+1;
            
        }
        
        return -1;
    }
};